Add ASCII opacity parser to mean_opacity_photons class.

RyanWollaeger opened the pull request:
+ Add ASCII grey data parser to MeanOpacity class in `mean_opacity_photons.hpp`
+ Combine Rosseland and Planck opacity into one `Databox`, with a 3rd discrete index valued:
  - 0 = Rosseland,
  - 1 = Planck.
+ Add python script for generating power-law opacity ASCII file, and example file for unit test.
+ Add unit test that parses toy opacity file (`kap_plaw.txt`).

RyanWollaeger committed:
Add Zhu dust+molecule+atom table opacity variant and example file.

+ Add ZhuTableOpacity photon class with Rosseland-Planck databox.
+ Assume log-log rho-T for DataBox interpolation.
+ Add ZhuTable variant to opac_photons.hpp.
+ Add unit test for ZhuTable variant.

Notes:
- See: Zhaohuan Zhu et al (2021).
- The next step is to convert the Zhu ASCII into HDF5 format(s).

RyanWollaeger commented:
@brryan @jonahm-LANL - can't seem to add reviewers - anyway here is a preliminary changeset - thanks for the help!

Yurlungur commented on singularity-opac/photons/zhu_data/opacitysolar09dustq3p5amax0p1new.txt:
not a huge fan of storing an ascii data file in the repo---but if this is the best solution we can work with it. It might be better to take a file name as input and distribute this file in a known place or have a script to fetch it. Same goes for the hdf5 binary. We could, if you like distribute them as artifacts in a data release via github's release mechanism. We do something similar for gold file data in singularity-eos.

RyanWollaeger replied:
@Yurlungur - thanks very much for the comments - @brryan also recommended removing the files, and instead using a smaller "toy" test file - will make that change.

RyanWollaeger committed:
Add option to parse HDF5 data in ZhuTable opacity constructor.

+ Use filesystem header to check extention in ctor.
+ Refactor old ASCII parser to a private function (called if .txt).
+ Add ZhuTable::Save method and use to add HDF5 file.
+ Update unit test to parse HDF5 data instead.

Thanks to Jonah Miller for the suggestion to use the Save method to
create the HDF5 file.

Yurlungur commented on singularity-opac/photons/zhu_grey_table_opacity_photons.hpp:
```diff
@@ -0,0 +1,309 @@
+
+  ZhuTableOpacity() = default;
+
+  // construct Planck/Rosseland DataBox from Zhu ascii file
```
```suggestion
  // construct Planck/Rosseland DataBox from Zhu ascii or spiner hdf5 file
```

Yurlungur commented on singularity-opac/photons/zhu_grey_table_opacity_photons.hpp:
```diff
@@ -0,0 +1,309 @@
+        // assume this is one of the original Zhu et al (2021) ASCII files
+        loadZhuASCII(ff);
+#ifdef SPINER_USE_HDF
+      } else if (extension == ".hdf5" || extension == ".h5") {
```
```suggestion
      } else if (extension == ".hdf5" || extension == ".h5" || extension == ".sp5") {
```

Yurlungur commented on singularity-opac/photons/zhu_grey_table_opacity_photons.hpp:
```diff
@@ -0,0 +1,309 @@
+
+  // construct Planck/Rosseland DataBox from Zhu ascii file
+  ZhuTableOpacity(const std::string filename, const bool use_planck_absorb = false)
+    : opac_type_(use_planck_absorb) {
```
Do you plan to load both types at once? It might be possible to generate them both with only one file load. Maybe like this:

```C++
ZhuTableOpacity(const ZhuTableOpacity &other, const bool use_planck_absorb = false)
    : dist_(other.dist_), kappa_(other.kappa_), opac_type_(use_planck_absorb) {}
```
so you can initialize one variant from the other.

RyanWollaeger replied:
Indeed, for use in transport specifically, was planning to permit even using both types at once in a hybrid-type scheme, which exists in the literature.  This constructor suggestion seems like a nice way to get both variants for instance, for grey diffusion. Will add it...

RyanWollaeger replied:
Following @brryan's suggestion, this entire class has been removed, and its contents moved into `mean_opacity_photons.hpp`. There the Rosseland and Planck values naturally get accessed separately.

Yurlungur replied:
ah that's a nice solution. :+1:

Yurlungur commented on test/CMakeLists.txt:
```diff
@@ -48,8 +48,17 @@ PRIVATE
+    ${CMAKE_CURRENT_BINARY_DIR}/opacitysolar09dustq3p5amax0p1new.txt SYMBOLIC)
+file(CREATE_LINK
+    ${CMAKE_SOURCE_DIR}/singularity-opac/photons/zhu_data/opacitysolar09dustq3p5amax0p1new.hdf5
+    ${CMAKE_CURRENT_BINARY_DIR}/opacitysolar09dustq3p5amax0p1new.hdf5 SYMBOLIC)
```
If we use a data release model, this would be replaced with cmake fetch content.

RyanWollaeger replied:
Modified this to use a smaller example .txt file, but left the symlink as-is in the most recent change.

Yurlungur replied:
the smaller file is certainly good enough IMO.

Yurlungur reviewed (approved):
Looks good! Some minor suggestions but overall this seems like a pretty clean implementation. My main suggestion/complaint is to move the data files out of git. If we want to distribute them, we could maybe use github's release machinery. Let's discuss when you have the chance.

RyanWollaeger committed:
Update singularity-opac/photons/zhu_grey_table_opacity_photons.hpp

Co-authored-by: Jonah Miller <DrLoco3000@gmail.com>

RyanWollaeger committed:
Add another valid HDF5 extension (thanks JMM).

Co-authored-by: Jonah Miller <DrLoco3000@gmail.com>

RyanWollaeger committed:
Put Zhu ASCII parsing into mean_opacity_photons (BRR suggestion).

+ Remove Zhu grey table class and unit test.
+ Replace Zhu data table with toy power-law data table.
+ Generalize file-based ctor to take ASCII or HDF5 file.
+ Add sub-test for parsing toy power-law data into MeanOpacity.

RyanWollaeger committed:
Combine Planck and Rosseland into one DataBox in MeanOpacity.

RyanWollaeger commented:
What do people think about slightly modifying the ASCII format: e.g. extending the header line with min/max density and temperature, so it can be parsed in rather than hard-coded?

Yurlungur commented:
> What do people think about slightly modifying the ASCII format: e.g. extending the header line with min/max density and temperature, so it can be parsed in rather than hard-coded?

I'm a bit nervous about doing that if the ASCII file is provided by someone else. However, if we assume control of it, that makes complete sense. You could also just demand the min/max densities be provided as inputs?

RyanWollaeger commented:
> > What do people think about slightly modifying the ASCII format: e.g. extending the header line with min/max density and temperature, so it can be parsed in rather than hard-coded?
> 
> I'm a bit nervous about doing that if the ASCII file is provided by someone else. However, if we assume control of it, that makes complete sense. You could also just demand the min/max densities be provided as inputs?

That's a fair point - right now, working on adding a preprocessing python script to generate a new file that: verifies the header data, adds the min/max rho-T values to the header calculated from the original file data, and optionally interpolates to a log-log grid in rho-T (if the grid is not log-log already).  This might help people vet their data, or provide a starting point.  However, we can defer modifying the ASCII that gets parsed in.

RyanWollaeger committed:
Add script to interpolate existing ASCII to log-log rho-T grid.

+ (Optionally) interpolate using SciPy interp2d.
+ Assert logarithmic grids if is_loglog is specified.
+ Verify number of rho-T points in header in old opacity file.
+ Add min/max rho-T as metadata to header in new opacity file.

Note: this script assumes old ASCII file is in Zhu et al (2021)
grey opacity format, which has a Rosseland and Planck column.

RyanWollaeger commented:
Just added the old->new ASCII preprocessing script in the last commit, but left the MeanOpacity class alone.

RyanWollaeger committed:
Modify ASCII parser to use min/max header data.

+ Make changes to parser name and comments to remove references.
+ Remove power-law opacity generation script.

Note: the new metadata is redundant with the rho-T column data,
so either might be removed.

RyanWollaeger committed:
Remove redundant rho-T columns from parsed ASCII format.

+ Remove rho-T parsing from MeanOpacity loadASCII function.
+ Update test ASCII data file example_ascii/kap_plaw.txt.

Note: non-uniform rho-T data points can be interpolated onto
a log-log grid, e.g., with example_ascii/preproc_ascii_opac.py.

RyanWollaeger commented:
> Just added the old->new ASCII preprocessing script in the last commit, but left the MeanOpacity class alone.

Now the `loadASCII` function is modified to take in a modified format, which has min/max rho-T in the header and no longer has rho-T columns. The preprocessing script has SciPy's 2D interpolation function, so can remap any non-uniformly spaced data that has rho and T columns to a log-log format, as @Yurlungur originally suggested.

Yurlungur commented:
I'm happy with this. @RyanWollaeger @brryan merge when you feel its ready.

brryan commented on singularity-opac/photons/example_ascii/kap_plaw.txt:
```diff
@@ -0,0 +1,19 @@
+ NRho 4 NT 4
+ rho_min 1.00000000e-14 rho_max 7.94328235e-01
```
Are these inclusive bounds, e.g. the first opacity value is at rho=1e-14 and the last is at rho=7.94...e-01, rather than the opacity values being "bin centers" where their `rho = exp(log(rho_min) + (i + 0.5)*dlogrho)` for the `i`th opacity value?

Would it be worth adding a short paragraph to the `README` explaining this format?

RyanWollaeger replied:
These are indeed inclusive bounds. Will add a description to the README.

RyanWollaeger replied:
Just added a section to the README.md file.

brryan replied:
Looks great thanks!

RyanWollaeger committed:
Add preliminary section on current ASCII capability to README.md.

brryan commented on singularity-opac/photons/example_ascii/preproc_ascii_opac.py:
```diff
@@ -0,0 +1,71 @@
+#--------------------------------------------------------------------------------------------------#
```
```suggestion
# ======================================================================
# © 2024. Triad National Security, LLC. All rights reserved.  This
# program was produced under U.S. Government contract
# 89233218CNA000001 for Los Alamos National Laboratory (LANL), which
# is operated by Triad National Security, LLC for the U.S.
# Department of Energy/National Nuclear Security Administration. All
# rights in the program are reserved by Triad National Security, LLC,
# and the U.S. Department of Energy/National Nuclear Security
# Administration. The Government is granted for itself and others
# acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
# license in this material to reproduce, prepare derivative works,
# distribute copies to the public, perform publicly and display
# publicly, and to permit others to do so.
# ======================================================================

#--------------------------------------------------------------------------------------------------#
```

brryan commented on singularity-opac/photons/example_ascii/preproc_ascii_opac.py:
```diff
@@ -0,0 +1,71 @@
+hdr = "NRho "+str(NRho)+" NT "+str(NT) + "\n" \
+    "rho_min {:.8e}".format(r[0])+" rho_max {:.8e}".format(r[NRho-1]) + "\n" \
+    "T_min {:.8e}".format(tt[0]) + " T_max {:.8e}".format(tt[NT-1])
+np.savetxt(args.fname_new, opac[:,2:], delimiter=" ", header=hdr, comments=" ", fmt="%.8e")
+
+#--------------------------------------------------------------------------------------------------#
+# End of preproc_ascii_opac.py
+#--------------------------------------------------------------------------------------------------#
```
```suggestion
np.savetxt(args.fname_new, opac[:,2:], delimiter=" ", header=hdr, comments=" ", fmt="%.8e")

```

brryan commented on singularity-opac/photons/mean_opacity_photons.hpp:
```diff
@@ -92,29 +113,27 @@ class MeanOpacity {
     Real lRho = toLog_(rho);
     Real lT = toLog_(temp);
-    return rho * fromLog_(lkappaPlanck_.interpToReal(lRho, lT));
+    return rho * fromLog_(lkappa_.interpToReal(lRho, lT, 1));
```
What do you think about adding a 
```
enum OpacityAveraging {
    Rosseland = 0,
    Planck = 1
};
```
so this can become `(lkappa_.interpToReal(lRho, lT, Planck)` here and elsewhere?

RyanWollaeger replied:
It's a very good idea - should this `enum` be added here, or in the `base` subdirectory?

brryan replied:
I would add it here inside the `MeanOpacity` class to avoid polluting the namespace

RyanWollaeger replied:
Just added the `enum` - thanks!

brryan replied:
:+1:

brryan reviewed (approved):
Few nitpicking comments, otherwise LGTM! Thanks for adding the test and documentation

RyanWollaeger committed:
Remove footer comment from ASCII preprocessing script.

Co-authored-by: Ben Ryan <brryan@lanl.gov>

RyanWollaeger committed:
Add copyright banner to ASCII preprocessing script.

Co-authored-by: Ben Ryan <brryan@lanl.gov>

RyanWollaeger committed:
Use MeanOpacity class member enum for Planck and Rosseland index.
(Thanks to BRR for the suggestion.)
